### ISSUE TITLE
Feynman slash notation is broken in HTML-CSS mode

### ISSUE THREAD

**Issue by bgromov** [mathjax/MathJax#2107]
### Issue Summary

Feynman slash notation (i.e. TeX `{a\!\!\!/}`) is broken in HTML-CSS, but not in other modes, e.g. SVG and Common HTML work fine.

![image](https://user-images.githubusercontent.com/3196635/50691990-b86bea00-1032-11e9-9af5-4bbfe9862c10.png)

### Steps to Reproduce

Minimal example for [MathJax live demo](https://www.mathjax.org/#demo):
```
${a\!\!\!/}{a\!\!\!/} \equiv a^{\mu }a_{\mu }\cdot I_{4}=a^{2}\cdot I_{4}$
```

In HTML-CSS mode symbols are rendered with a wrong offset:

![image](https://user-images.githubusercontent.com/3196635/50692710-0eda2800-1035-11e9-8053-635e9a9b90ef.png)

but in, e.g. SVG mode, everything works fine:

![image](https://user-images.githubusercontent.com/3196635/50692763-35985e80-1035-11e9-8f4b-73e8ecdf5cae.png)

More equations can be found here: https://en.wikipedia.org/wiki/Feynman_slash_notation

### Technical details:

* MathJax Version: 2.7.5
* Client OS: macOS Mojave 10.14.2 (18C54)
* Browser: Chrome 70.0.3538.110 (Official Build) (64-bit)

Thanks,

Boris

**Comment by dpvc** [comment]
The issue is a bit more complicated than it may seem at first.  The expression `{a\!\!\!/}` is using spacing hacks to overlay one character on another.  Such expressions are very sensitive to the metrics of the font in use, and are not reliable in general, and so are in appropriate for environments where the fonts is not tightly controlled.

MathJax's HTML-CSS output renderer can use more than one font, and it checks your system to see if you have one that it knows about installed locally (so that it doesn't have to download it over the web).  It appears from your screen shots that you have the STIX fonts installed locally, as that is the font being used, rather than MathJax's TeX font.

The TeX fonts are constructed very carefully to interoperate in particular ways (like the slash `/` being exactly three `\!` wide).  Unfortunately, the STIX fonts don't have those same metrics, and the slash is much thinner than that.  So when you backspace using three `\!`, that is too far for the spacing of the slash in the STIX fonts.  That is why you are getting the strange spacing in your expressions (you are backing up too far and so things that follow are overlapping badly).

You do not see this in CommonHTML and SVG output because those do not use locally installed fonts, so they will not switch to STIX fonts but remain with the default MathJax TeX fonts, where your backspacing has the correct size.

Backspacing like this is rarely a good idea.  It would be better to use `\llap{}`, as in `a\llap{/}`, which would at least make sure that what follows doesn't overlap the `a`, as is currently happening for you.  On the other hand, the STIX version of `/` is not centered vertically on the math axis, as is the one in the MathJax TeX fonts, and so the result is still not very satisfying.  It would be better to use the unicode character at U+2215 (fraction slash) than the text solidus at U+002F.  You can get that using `a\llap{\unicode{x2215}}`.  This will work reasonably well in both STIX and the MathJax TeX fonts.

If you control the web site, you can arrange for the HTML-CSS output jax to not switch to STIX fonts.  Alternatively, you can install the MathJax TeX fonts locally on your computer (as it will use local MathJax TeX fonts in preference to local STIX fonts).

**Comment by FranklinYu** [comment]
>  The expression `{a\!\!\!/}` is using spacing hacks to overlay one character on another.

Is there a non-hack way to overlay one character on another?

**Comment by dpvc** [comment]
In general, it doesn't work well regardless.  There are characters designed for doing it, e.g., the combining characters like those in the Combining Diacritical Marks block of unicode (starting at U+0300).  The combining long solidus overlay (U+0338) is one that puts a slash over a character, but it is designed to go over relations (like `=` or `>`), and those are a slight bit larger than most letters, so the placement is not perfect for letters.  And since some letters are wider than others (e.g., `m` compared to `i`), the placement is better for some than for others.  You could use `m\unicode{x338}`, but `i\unicode{x338}` would be less successful.

The `\llap{}` approach (or `\rlap{}` in some cases) is another way to do this, but it suffers the same sort of problems.  `a\llap{\unicode{x2215}}`, as I suggested above, is a reasonable way to do this without resorting to spacing hacks.

The real problem is that there is no natural way to center one thing on another in MathML, which is the format that underlies MathJax.  (MathML doesn't have the idea of stretchable glue that TeX uses for this.)  There are hacks that can be used, but they are pretty awful.  One is to turn the base letter into an operator (like summation sign) and put the seas as a limit above (that gets centered automatically), but to lower the slash so that it is actually on top of the letter.  Of course, that requires knowing the height of the letter, but you can use a `\strut` to force the height to be a common one for all letters.  You have to use `\smash` to remove the depth of the lowered slash from being taken into account.  The result looks right, but it actually has extra height because of the operator-with-limits construct.  So you need to `\smash` the result and use a `\strut` to give it the a reasonable height (basically the height of the slash).

A definition makes it easier to use:

```
\def\fslash#1{\mathord{\smash{\mathop{#1\strut}\limits^{\smash{\textstyle\lower10pt{\unicode{x2215}}}}}\strut}}
```

Then `\fslash{a}` will center the slash over the `a` no matter what the widths of the characters in the font in use.  (the only metric you need to know is how far to drop the character in relation to the `\strut` (here I use 10pt because the strut is 8.6pt plus a little for the space between the base and the limits above).

This will center properly over any width character (e.g., `m` and `i`), and even over multi-character bases, or things like `\longrightarrow`.  It works in all the output modes, and all the fonts.  Because of the `\textstyle`, though, it will not work properly in super- or sub-scripts (one could probably use `\mathchoice` to resolve that, but it would take some work).

This is a horrible hack, however, and the results are completely non-semantic.  I don't recommend you use it.  

PS, I see you have modified your answer while I was writing this one.  The definition I gave above is the sort of thing that is being recommended on the page you found (particularly the solution using `\centernot`, which MathJax doesn't implement).

**Comment by alexchandel** [comment]
Part of the frequency of this issue is Wikipedia has no LaTeX package whatsoever to overlay characters or strikethrough individual characters.  `\smash`, `\rlap`, `\llap`, etc are all unimplemented, resulting in hacks for slashes and strikethroughs littering Wikipedia.

**Comment by dpvc** [comment]
An alternative that works reasonably well in all the MathJax fonts (except NeoEuler) is `\unicode{x2215}\kern-.5em a`.  You can adjust the kern for wider or thinner letters, if needed.  Of course, getting people on Wikipedia to use this is probably a lost cause.  Without `\llap`, `\rlap` and `\smash`, I don't see any way of handling this in a font-agnostic way.
